lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java

```diff
@@ -0,0 +1,123 @@
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.lock.impl.log.Log;
+
+/**
+ * This class holds the logic to schedule/abort requests that need to be completed at a given time.
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @since 9.2
+ */
+public class RequestExpirationScheduler {
+
+   private static final Log log = LogFactory.getLog(RequestExpirationScheduler.class, Log.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private final ScheduledExecutorService scheduledExecutorService;
+   private final ConcurrentHashMap<String, ScheduledRequest> scheduledRequests = new ConcurrentHashMap<>();
+
+   public RequestExpirationScheduler(ScheduledExecutorService scheduledExecutorService) {
+      this.scheduledExecutorService = scheduledExecutorService;
+   }
+
+   class ScheduledRequest {
+      private String requestId;
+      private CompletableFuture<Boolean> request;
+      private ScheduledFuture<?> scheduledFuture;
+
+      public ScheduledRequest(String requestId, CompletableFuture<Boolean> request, ScheduledFuture<?> scheduledFuture) {
+         this.requestId = requestId;
+         this.request = request;
+         this.scheduledFuture = scheduledFuture;
+      }
+
+      public String getRequestId() {
+         return requestId;
+      }
+
+      public CompletableFuture<Boolean> getRequest() {
+         return request;
+      }
+
+      public ScheduledFuture<?> getScheduledFuture() {
+         return scheduledFuture;
+      }
+   }
+
+   /**
+    * Schedules a request for completion
+    *
+    * @param requestId, the unique identifier if the request
+    * @param request, the request
+    * @param time, time expressed in long
+    * @param unit, {@link TimeUnit}
+    */
+   public void scheduleForCompletion(String requestId, CompletableFuture<Boolean> request, long time, TimeUnit unit) {
+      if (request.isDone()) {
+         if(trace) {
+            log.tracef("Request[%s] is not scheduled because is already done", requestId);
+         }
+         return;
+      }
+
+      if (scheduledRequests.containsKey(requestId)) {
+         if(trace) {
+            log.tracef("Request[%s] is not scheduled because it is already scheduled", requestId);
+         }
+         return;
+      }
+
+      if(trace) {
+         log.tracef("Request[%s] being scheduled to be completed in [%d, %s]", requestId, time, unit);
+      }
+
+      ScheduledFuture<?> scheduledFuture = scheduledExecutorService.schedule(() -> {
+         request.complete(false);
+         scheduledRequests.remove(requestId);
+      }, time, unit);
+
+      scheduledRequests.putIfAbsent(requestId, new ScheduledRequest(requestId, request, scheduledFuture));
```

Comment:
Hrmm this seems odd to me. I wonder what should we do if we have multiple requests for the same request Id ? I would think we should probably set one to exception maybe?

Reply:
we should't have more than one requestId, this should be unique. if not, it's a bug for me

Reply:
Then we should probably throw an exception or set an exception on a future or something here.